core/jobs/batch_jobs/exp_migration_jobs.py

```diff
@@ -0,0 +1,470 @@
+                self._generate_exploration_changes)
+        )
+
+        exp_objects_list = (
+            {
+                'exp_model': unmigrated_exploration_models,
+                'exp_summary_model': exp_summary_models,
+                'exp_rights_model': exp_rights_models,
+                'exploration': migrated_exp,
+                'exp_changes': exp_changes
+            }
+            | 'Merge objects' >> beam.CoGroupByKey()
+            | 'Get rid of ID' >> beam.Values() # pylint: disable=no-value-for-parameter
+            | 'Remove unmigrated explorations' >> beam.Filter(
+                lambda x: len(x['exp_changes']) > 0
+                    and len(x['exploration']) > 0)
+            | 'Reorganize the exploration objects' >> beam.Map(lambda objects: {
+                'exp_model': objects['exp_model'][0],
+                'exploration': objects['exploration'][0],
+                'exp_rights_model': objects['exp_rights_model'][0],
+                'exp_summary_model': objects['exp_summary_model'][0],
+                'exp_changes': objects['exp_changes']
+            })
+        )
+
+        exp_objects_list_job_run_results = (
+            exp_objects_list
+            | 'Transform exp objects into job run results' >> (
+                job_result_transforms.CountObjectsToJobRunResult(
+                    'EXP MIGRATED'))
+        )
+
+        cache_deletion_job_run_results = (
+            exp_objects_list
+            | 'Delete exploration from cache' >> beam.Map(
+                lambda exp_object: self._delete_exploration_from_cache(
+                    exp_object['exploration']))
+            | 'Generate results for cache deletion' >> (
+                job_result_transforms.ResultsToJobRunResults('CACHE DELETION'))
+        )
+
+        exp_models_to_put = (
+            exp_objects_list
+            | 'Generate exploration models to put' >> beam.FlatMap(
+                lambda exp_objects: self._update_exploration(
+                    exp_objects['exp_model'],
+                    exp_objects['exp_rights_model'],
+                    exp_objects['exploration'],
+                    exp_objects['exp_changes'],
+                ))
+        )
+
+        exp_summary_models_to_put = (
+            exp_objects_list
+            | 'Generate exp summary models to put' >> beam.Map(
+                lambda exp_objects: self._update_exp_summary(
+                    exp_objects['exploration'],
+                    exp_objects['exp_summary_model'],
+                    exp_objects['exp_rights_model']
+                ))
+        )
+
+        exp_opp_summary_models_to_put = (
+            migrated_exp
+            | 'Generate exp opportunity summary models to put' >> (
+                beam.MapTuple(
+                    self._update_exploration_opportunity_summary_models,
+                    exp_id_to_exp_opp_summary_model=(
+                        exp_opp_summary_models_to_exp)))
+        )
```

Comment:
Can we do this together with rest of the stuff using `CoGroupByKey` above? So we first tie the models together into one dict and then update the opportunity?